tokens-accounting/src/cache.rs

```diff
@@ -0,0 +1,229 @@
+        match self.data.token_data.get_data(id) {
+            accounting::GetDataResult::Present(d) => Ok(Some(d.clone())),
+            accounting::GetDataResult::Deleted => Ok(None),
+            accounting::GetDataResult::Missing => {
+                Ok(self.parent.get_token_data(id).map_err(|_| Error::ViewFail)?)
+            }
+        }
+    }
+
+    fn get_circulating_supply(&self, id: &TokenId) -> Result<Option<Amount>, Self::Error> {
+        let parent_supply = self.parent.get_circulating_supply(id).map_err(|_| Error::ViewFail)?;
+        let local_delta = self.data.circulating_supply.data().get(id).cloned();
+        combine_amount_delta(&parent_supply, &local_delta).map_err(Error::AccountingError)
+    }
+}
+
+impl<P> FlushableTokensAccountingView for TokensAccountingCache<P> {
+    type Error = Error;
+
+    fn batch_write_tokens_data(
+        &mut self,
+        delta: TokensAccountingDeltaData,
+    ) -> Result<crate::data::TokensAccountingDeltaUndoData, Self::Error> {
+        self.data.merge_with_delta(delta)
+    }
+}
+
+impl<P: TokensAccountingView> TokensAccountingOperations for TokensAccountingCache<P> {
+    fn issue_token(&mut self, id: TokenId, data: TokenData) -> Result<TokenAccountingUndo, Error> {
+        if self.get_token_data(&id)?.is_some() {
+            return Err(Error::TokenAlreadyExists(id));
+        }
+
+        if self.get_circulating_supply(&id)?.is_some() {
+            return Err(Error::TokenAlreadyExists(id));
+        }
+
+        let undo_data = self
+            .data
+            .token_data
+            .merge_delta_data_element(id, accounting::DataDelta::new(None, Some(data)))?;
+
+        Ok(TokenAccountingUndo::IssueToken(IssueTokenUndo {
+            id,
+            undo_data,
+        }))
+    }
+
+    fn mint_tokens(
+        &mut self,
+        id: TokenId,
+        amount_to_add: Amount,
+    ) -> Result<TokenAccountingUndo, Error> {
+        let token_data = self.get_token_data(&id)?.ok_or(Error::TokenDataNotFound(id))?;
+        let circulating_supply = self.get_circulating_supply(&id)?.unwrap_or(Amount::ZERO);
+
+        match token_data {
+            TokenData::FungibleToken(data) => match data.total_supply() {
+                TokenTotalSupply::Fixed(limit) => {
+                    let expected_circulating_supply =
+                        (circulating_supply + amount_to_add).ok_or(Error::AmountOverflow)?;
+                    if expected_circulating_supply > *limit {
+                        return Err(Error::MintExceedsSupplyLimit(amount_to_add, *limit, id));
+                    }
+                }
+                TokenTotalSupply::Lockable => {
+                    if data.is_locked() {
+                        return Err(Error::CannotMintFromLockedSupply(id));
+                    }
+                }
+                TokenTotalSupply::Unlimited => { /* do nothing */ }
+            },
+        };
+
+        self.data.circulating_supply.add_unsigned(id, amount_to_add)?;
```

Comment:
I don't quite understand what happens when some amount is added to the circulating supply. How do we get it from the circulating supply account to a utxo? Conversely, what happens when you unmint? How are the tokens removed from circulation?

Reply:
When you mint tokens by spending an input from Account::TotalSupply you are allowed to produce a utxo with amount <= the one specified for input.
So a reference tx for minting looks like this:  `Account(TotalSupply) + Transfer(Coins) -> Transfer(Tokens) + Burn(Coins)`

When unminting you spend from an account input to authorize the operation and also you burn utxos: `Account(CirculatingSupply) + Transfer(Tokens) + Transfer(Coins) -> Burn(Tokens) + Burn(Coins)`
The amount of tokens that was burned is the amount that is actually taken out of circulation in the db.

Reply:
What are the invariants here? Something like like the following?

* sum of all utxos for given token = circulating supply
* curculating supply < total supply

I have to admit I only skimmed the tests. Is there some test, preferably randomised, that checks the invariants after some arbitrary operations?

Reply:
the invariant are these:
- sum of all spends from Account::TotalSupply == circulating supply
- sum of all utxos <= circulating supply (because we allow outputs to be less then inputs so tokens can theoretically go to miner or not been produced for some reason)
- circulating supply <= total supply

Right now there are only deterministic tests and I've been working on a randomised in a separate branch